### PR TITLE
Add docs for PII management

### PR DESCRIPTION
**Pull request by ancalita**
**Proposed changes**:
- ...

**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [x] updated the documentation
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [ ] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

File: docs/docs/pii-management.mdx [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+
+You can now anonymize personally identifiable information (PII) data in your logs and events streamed via Kafka event
+broker only. Continue reading to learn how to enable the anonymization of PII data.
+
+:::
+
+Management of sensitive customer data collected by assistants is a critical requirement for complying with regulations
+and managing the data securely while making it available for analysis.
+Rasa Pro 3.6 introduces the ability to anonymize PII data in your logs and events streamed via the Kafka event broker.
+Note that the tracker store continues to store conversation data which is not anonymized because the tracker store data
```
Comment (andreaklmn) [inline docs/docs/pii-management.mdx:25]: Can we make it more clear here why we are not anonimizing the tracker store in this version? 
Eg.: we want to retain a source of truth for pristine data for non-repudiation purposes. And probably also prediction..? (not sure)

File: docs/docs/pii-management.mdx [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+plays a vital role in the assistant's dialogue management.
+
+## Architecture Overview
+
+The anonymization of the Rasa event is run through an anonymization pipeline at the end of the dialogue management
```
Comment (andreaklmn) [inline docs/docs/pii-management.mdx:30]: such a great overview:)

**Review by andreaklmn — CHANGES_REQUESTED** [review]
added a tiny comment:)

File: docs/docs/pii-management.mdx (hunk 1 above)
Comment (ancalita) [reply docs/docs/pii-management.mdx]: Addressed 👍🏻 
https://github.com/RasaHQ/rasa/pull/12505/commits/f1b891cdcec2108822c99ebc3dc629c88d44c5b4

File: docs/docs/pii-management.mdx [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+
+Note that the tracker store continues to store conversation data which is not anonymized. This is required so that we
+retain a source of truth for pristine data required for non-repudiation purposes. In addition, because the tracker store
+data plays a vital role in the assistant's dialogue management at inference time, anonymizing data in flight could have
+undesirable consequences for the assistant's action predictions.
+
+## Architecture Overview
+
+The anonymization of the Rasa event is run through an anonymization pipeline at the end of the dialogue management
+action prediction and execution. The Rasa agent will invoke the anonymization pipeline during each user message
```
Comment (souvikg10) [inline docs/docs/pii-management.mdx:34]: > The Rasa agent will invoke the anonymization pipeline during each user message handling before saving the un-anonymized new events to the tracker store.

This line sounds a little contradicting 😕 we are running the anonymization pipeline and then saving un-anonymized events to the tracker store?

File: docs/docs/pii-management.mdx [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+  client_id: kafka-python-rasa
+```
+
+The `anonymization_topics` section contains a list of Kafka topics to which the anonymized events will be published.
+Each Kafka topic must have a `name` field and an `anonymization_rules` field. The `name` field specifies the name of the
+Kafka topic. The `anonymization_rules` field specifies the `id` of the anonymization rule list to be used for the Kafka topic.
+
+## How to enable anonymization of PII in logs
+
+You can enable anonymization of PII in logs by filling the `logger` section in the `endpoints.yml` file.
```
Comment (souvikg10) [inline docs/docs/pii-management.mdx:175]: How is this different from `custom-logging-configuration`?

File: docs/docs/pii-management.mdx [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+The `logger` section must have the following structure:
+
+```yaml
+logger:
+  formatter:
+    anonymization_rules: rules_1
```
Comment (souvikg10) [inline docs/docs/pii-management.mdx:181]: can this be also a list of rules?

Comment (ancalita) [reply docs/docs/pii-management.mdx:196]: No, this can only be a string and must match and `id` of the anonymization rule list in the `anonymization` section. So all rules are grouped only in that section.

File: docs/docs/pii-management.mdx (hunk 3 above)
Comment (ancalita) [reply docs/docs/pii-management.mdx:34]: It's not contradictory, I mention in L26-29 that the tracker store remains single source of truth for pristine unanonymized data. The pipeline only anonymizes data streamed to specific topics consumed by Kafka consumers. Let me know how I should rephrase this to be clear.

File: docs/docs/pii-management.mdx (hunk 4 above)
Comment (ancalita) [reply docs/docs/pii-management.mdx:190]: The logging file CLI option is used for configuring the  `logging` Python std library which was the dominant logging method we used in the codebase. However, in order to be able to hook the pipeline in PII leaking logs only (and not across all the possible logs), we chose to use `structlog` library instead which uses the [concept of processor chains](https://www.structlog.org/en/stable/getting-started.html#manipulating-log-entries-in-flight), we created our own processor (the anonymizer). In a nutshell, you can't use the logging file with structlog as far as I'm aware or it's for a separate ticket to work on how to enable this. Let me know if any of this needs to be clarified in the docs. I thought it's an implementation detail but the reality is that now the logs will be a mix of new structured logs and old logs with text messages.

File: docs/docs/pii-management.mdx [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+
+Management of sensitive customer data collected by assistants is a critical requirement for complying with regulations
+and managing the data securely while making it available for analysis.
+Rasa Pro 3.6 introduces the ability to anonymize PII data in your logs and events streamed via the Kafka event broker.
+
+Note that the tracker store continues to store conversation data which is not anonymized. This is required so that we
+retain a source of truth for pristine data required for non-repudiation purposes. In addition, because the tracker store
+data plays a vital role in the assistant's dialogue management at inference time, anonymizing data in flight could have
+undesirable consequences for the assistant's action predictions.
+
```
[the remark this answers is not in the document]
Comment (ancalita) [reply docs/docs/pii-management.mdx:32]: Addressed: https://github.com/RasaHQ/rasa/pull/12505/commits/89614a3ac29e11cec847a8e67ccc3a0407aad78f
I've clarified a few other assumptions and added some more detail in other parts, so re-requesting review.

File: docs/docs/glossary.mdx [hunk 7]
```diff
@@ -17,6 +17,11 @@ The server that runs custom action code, separate from Rasa. Rasa maintains the
 
 Adding labels to messages and conversations so that they can be used to train a model.
 
+## [Anonymization](./pii-management.mdx)
+
+The process of replacing personally identifiable information (PII) with masked, artificial or constant text values.
```
Comment (ancalita) [inline docs/docs/glossary.mdx:22]: I saw a review note on this in my email notifications, but it doesn't seem to be in the GH UI anymore.
In any case, could you please pseudo-anonymization? Is that the same as replacing real PII with fake generated data? In which case I mention here as `artificial`, so I consider this to be a subset of anonymization. Let me know how I should rephrase to make it clearer.

File: docs/docs/pii-management.mdx (hunk 4 above)
Comment (souvikg10) [reply docs/docs/pii-management.mdx:192]: I am not sure how many are using custom logging configuration anyway, perhaps we can move custom logging configurations to rasa pro and use endpoints.yml to harmonize it. but nevertheless it is a separate conversation. I dont see it contradicting the docs here so we are good

File: docs/docs/pii-management.mdx (hunk 3 above)
Comment (souvikg10) [reply docs/docs/pii-management.mdx:36]: can we be explicit about the sequence of events that happens here.
- Agent invokes the anonymization pipeline
- Anonymized events are published onto the event broker
- un-anonymized events are then saved in the tracker store.

In the architecture overview, i think have the order of what happens to the events is important. 
Also are we only anonymising the UserMessage event or also BotUttered?

Comment (ancalita) [reply docs/docs/pii-management.mdx]: > can we be explicit about the sequence of events that happens here

Addressed: can we be explicit about the sequence of events that happens here

> are we only anonymising the UserMessage event or also BotUttered?

We are anonymizing UserUttered, BotUttered, SlotSet and EntitiesAdded, I've clarified the events in a subsection just below the Architecture Overview, although used their class type name rather than code reference name. Let me know which one would be clearer.

Comment (souvikg10) [reply docs/docs/pii-management.mdx]: I think class type is clear. makes sense. thank you. everything looks very clear to me 💯

Comment (ancalita) [reply docs/docs/pii-management.mdx]: Yay, that's great! Thank you for your thorough review ⚡ I will wait for Lena's review on Monday before merging this.

**Review by l2nguyen — APPROVED** [review]

File: docs/docs/command-line-interface.mdx [hunk 8]
```diff
@@ -49,6 +49,8 @@ In addition to CLI arguments, several environment variables allow you to control
 2. LOG_LEVEL_MATPLOTLIB: This is the specialized environment variable to configure log level only for Matplotlib.
 3. LOG_LEVEL_RABBITMQ: This is the specialized environment variable to configure log level only for AMQP libraries, at the moment it handles log levels from `aio_pika` and `aiormq`.
 4. LOG_LEVEL_KAFKA: This is the specialized environment variable to configure log level only for kafka.
+5. LOG_LEVEL_PRESIDIO: This is the specialized environment variable to configure log level only for presidio, at the moment it handles log levels from `presidio_analyzer` and `presidio_anonymizer`.
```
[the remark this answers is not in the document]
Comment (ancalita) [reply docs/docs/command-line-interface.mdx:52]: Nope, you can't set the log level separately for the two dependencies, both are covered by this env var.

File: docs/docs/pii-management.mdx [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+:::note
+Not all languages have a pre-trained language model available. If you want to use a language that does not have
+a pre-trained language model available, you must train your own [spaCy](https://spacy.io/usage/training), [stanza](https://stanfordnlp.github.io/stanza/training.html)
+or [huggingface](https://huggingface.co/docs/transformers/training) model and install it in your Rasa Pro environment.
+:::
+
+### How to populate the rule_lists section
+
+The `rule_lists` section contains a list of anonymization rule lists. Each rule list must have a unique `id` of type
+string and a list of `rules`. Each rule must have an `entity` field and a `substitution` field. The `entity` field
```
[the remark this answers is not in the document]
Comment (ancalita) [reply docs/docs/pii-management.mdx:145]: Nope, not possible at the moment, we only have support for the built-in Presidio entities. I'll see how I can clarify that regex is not supported.

File: docs/docs/pii-management.mdx [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+       transformers: dslim/bert-base-NER
+    model_provider: transformers
+```
+#### How to install the language  model
+
+You must install the model name you declare to the `model_name` field in your Rasa Pro environment.
+For example, if you declare `model_name: en_core_web_lg`, you must install the spaCy `en_core_web_lg` model in your Rasa Pro environment.
+You can follow model installation instructions for spaCy and stanza models in the [Presidio official documentation](https://microsoft.github.io/presidio/analyzer/nlp_engines/spacy_stanza/).
+
+In the case of the `transformers` model provider, you must install the model name you declare to the `model_name` field
```
[the remark this answers is not in the document]
Comment (ancalita) [reply docs/docs/pii-management.mdx:131]: Good catch!

File: docs/docs/pii-management.mdx [hunk 11]
```diff
@@ -0,0 +1,213 @@
...
+:::
+
+### How to populate the rule_lists section
+
+The `rule_lists` section contains a list of anonymization rule lists. Each rule list must have a unique `id` of type
+string and a list of `rules`. Each rule must have an `entity` field and a `substitution` field. The `entity` field
+specifies the Presidio entity type to be anonymized and must be in uppercase.
+
+The `substitution` field specifies the anonymization method to be used. Currently, the
+following anonymization methods are supported: `text`, `mask`, and `faker`. If no substitution method is specified,
```
[the remark this answers is not in the document]
Comment (ancalita) [reply docs/docs/pii-management.mdx:149]: Good point 👍🏻

File: docs/docs/pii-management.mdx (hunk 9 above)
[the remark this answers is not in the document]
Comment (ancalita) [reply docs/docs/pii-management.mdx:145]: Addressed: https://github.com/RasaHQ/rasa/pull/12505/commits/ae10dcef4e1c8efe17e3e4f37b888af4506d3bd8

File: docs/docs/command-line-interface.mdx (hunk 8 above)
Comment (andreaklmn) [inline docs/docs/command-line-interface.mdx:52]:
```suggestion
5. LOG_LEVEL_PRESIDIO: This is the specialized environment variable to configure log level only for Presidio, at the moment it handles log levels from `presidio_analyzer` and `presidio_anonymizer`.
```

File: docs/docs/pii-management.mdx [hunk 12]
```diff
@@ -0,0 +1,259 @@
...
+broker only. Continue reading to learn how to enable the anonymization of PII data.
+
+:::
+
+Management of sensitive customer data collected by assistants is a critical requirement for complying with regulations
+and managing the data securely while making it available for analysis.
+Rasa Pro 3.6 introduces the ability to anonymize PII data in your logs and events streamed via the Kafka event broker.
+
+Note that the Tracker Store continues to store conversation data which is not anonymized. This is required so that we
+retain a source of truth for pristine data required for non-repudiation purposes. Learn how to secure your Tracker Store
```
Comment (andreaklmn) [inline docs/docs/pii-management.mdx:27]:
```suggestion
retain a source of truth of pristine data for non-repudiation purposes. Learn how to secure your Tracker Store
```

File: docs/docs/pii-management.mdx [hunk 13]
```diff
@@ -0,0 +1,259 @@
...
+anonymizing data in flight could have undesirable consequences for the assistant's dialogue management action predictions.
+
+## Architecture Overview
+
+The anonymization of the Rasa event is run through an anonymization pipeline at the end of the dialogue management
+action prediction and execution. Note that the processing done by the anonymization pipeline is scheduled as a background
+task and does not affect the assistant's response time.
+
+The anonymization steps are as follows:
+1. The Rasa agent will invoke the anonymization pipeline during each user message handling.
```
Comment (andreaklmn) [inline docs/docs/pii-management.mdx:40]:
```suggestion
1. The Rasa agent invokes the anonymization pipeline during each user message handling.
```

File: docs/docs/pii-management.mdx [hunk 14]
```diff
@@ -0,0 +1,259 @@
...
+2. The anonymization pipeline will run through a sequence of anonymization rules that are applied to the new events.
```
Comment (andreaklmn) [inline docs/docs/pii-management.mdx:41]:
```suggestion
2. The anonymization pipeline runs through a sequence of anonymization rules that are applied to the new events.
```

File: docs/docs/pii-management.mdx [hunk 15]
```diff
@@ -0,0 +1,259 @@
...
+3. The pipeline will publish the anonymized events to the Kafka topic that is mapped to the anonymization rule list in
```
Comment (andreaklmn) [inline docs/docs/pii-management.mdx:42]:
```suggestion
3. The pipeline publishes the anonymized events to the Kafka topic that are mapped to the anonymization rule list in
```

File: docs/docs/pii-management.mdx [hunk 16]
```diff
@@ -0,0 +1,259 @@
...
+`endpoints.yml`.
+4. The tracker store saves the original events which are not anonymized.
```
Comment (andreaklmn) [inline docs/docs/pii-management.mdx:44]:
```suggestion
4. The tracker store saves the original events that are not anonymized.
```

File: docs/docs/pii-management.mdx [hunk 17]
```diff
@@ -0,0 +1,259 @@
...
+        - entity: IBAN_CODE
+          substitution: mask
+```
+
+### How to populate the metadata section
+
+The metadata section contains the following fields: `language`, `model_name`, and `model_provider`.
+
+The `language` field specifies the language code of the text to be anonymized. Note that you can only specify one language
+per anonymization pipeline because the assumption made was that an assistant will be developed for a single language.
```
Comment (andreaklmn) [inline docs/docs/pii-management.mdx:99]:
```suggestion
per anonymization pipeline.
```

Comment (andreaklmn) [reply docs/docs/pii-management.mdx:99]: I deleted this because it made us look like we are building explicitly for single language bots. But perhaps if we phrased it differently?

File: docs/docs/pii-management.mdx [hunk 18]
```diff
@@ -0,0 +1,259 @@
...
+If you want to use `spacy`, we strongly recommend to use the available large models, such as `en_core_web_lg` or `es_core_news_lg`.
+This is to ensure more accurate entity recognition.
+
+:::caution
+When using large models, you must ensure that your Rasa Pro environment has enough memory to load the model.
+
+:::
+
+If you opt for using the `transformers` model provider, you must specify two model names in the `model_name` field:
+the HuggingFace model name and the spaCy model name, where the spacy model would wrap the transformers NER model.
```
Comment (andreaklmn) [inline docs/docs/pii-management.mdx:113]:
```suggestion
the HuggingFace model name and the spaCy model name, where the spaCy model would wrap the transformers NER model.
```

**Review by andreaklmn — CHANGES_REQUESTED** [review]

File: docs/docs/pii-management.mdx (hunk 17 above)
Comment (ancalita) [reply docs/docs/pii-management.mdx:99]: We need to document this assumption because the pipeline won't be able to work with language switching at the moment. How about: `because the capability is currently not able to handle language switching by end users.`

Comment (ancalita) [reply docs/docs/pii-management.mdx]: Or maybe replace `because` with `therefore` 😄

**Review by andreaklmn — APPROVED** [review]
